h2o-core/src/main/java/water/fvec/persist/FramePersist.java

```diff
@@ -0,0 +1,189 @@
+
+        private final Job<Frame> job;
+        private final Frame frame;
+        private final URI metaUri;
+
+        public SaveFrameDriver(
+            Job<Frame> job, Frame frame,
+            URI metaUri
+        ) {
+            this.job = job;
+            this.frame = frame;
+            this.metaUri = metaUri;
+        }
+
+        @Override
+        public void compute2() {
+            frame.read_lock(job._key);
+            new SaveChunksTask(metaUri.toString()).doAll(frame).join();
+            tryComplete();
+        }
+
+        @Override
+        public void onCompletion(CountedCompleter caller) {
+            frame.unlock(job);
+        }
+
+        @Override
+        public boolean onExceptionalCompletion(Throwable t, CountedCompleter caller) {
+            frame.unlock(job);
+            return super.onExceptionalCompletion(t, caller);
+        }
+    }
+
+    static class SaveChunksTask extends MRTask<SaveChunksTask> {
+
+        private final String metaUri;
+
+        SaveChunksTask(String metaUri) {
+            this.metaUri = metaUri;
+        }
+
+        @Override
+        public void map(Chunk[] cs) {
+            PersistUtils.write(getDataUri(metaUri, cs[0].cidx()), ab -> writeChunks(ab, cs));
+        }
+
+        private void writeChunks(AutoBuffer autoBuffer, Chunk[] chunks) {
+            for (Chunk c : chunks) {
+                autoBuffer.put(c);
+            }
+        }
+    }
+
+    public static Job<Frame> loadFrom(Key<Frame> key, String uri) {
+        URI metaUri = getMetaUri(key, sanitizeUri(uri));
+        FrameMeta meta = read(metaUri, AutoBuffer::get);
+        if (meta.numNodes != H2O.CLOUD.size()) {
+            throw new IllegalArgumentException("To load this frame a cluster with " + meta.numNodes + " nodes is needed.");
+        }
+        Job<Frame> job = new Job<>(meta.key, "water.fvec.Frame", "Load frame");
+        return job.start(new LoadFrameDriver(metaUri.toString(), meta), 1);
+    }
+
+    public static class LoadFrameDriver extends H2O.H2OCountedCompleter<LoadFrameDriver> {
+
+        private final String metaUri;
+        private final FrameMeta meta;
+
+        public LoadFrameDriver(
+            String metaUri, FrameMeta meta
+        ) {
+            this.metaUri = metaUri;
+            this.meta = meta;
+        }
+
+        @Override
+        public void compute2() {
+            Vec con = null;
+            try {
+                long nrow = meta.espc[meta.espc.length-1];
+                int nchunk = meta.espc.length-1;
+                con = Vec.makeConN(nrow, nchunk);
+                new LoadChunksTask(metaUri, meta.vecs).doAll(con).join();
+            } finally {
+                if (con != null) con.remove();
+            }
+            Futures fs = new Futures();
+            int rowLayout = Vec.ESPC.rowLayout(meta.vecs[0]._key, meta.espc);
+            for (Vec v : meta.vecs) {
+                v._rowLayout = rowLayout;
```

Comment:
This could be imho encapsulated in the FrameMeta class. (writeAll_impl, readAll_impl)

Reply:
writeAll_impl, readAll_impl are specific to Keyed, FrameMeta is not keyed and I dont want to put it into DKV which AutoBuffer.getKey does, I could add method to FrameMeta that would do the ab.get() and then do the rowLayout post-processing of loaded vecs but that adds another for-loop iterating through all the vecs and the code does not seem cleaner, WDYT?